src/api/c/fftconvolve.cpp

```diff
@@ -121,63 +134,70 @@ AF_BATCH_KIND identifyBatchKind(const dim4 &sDims, const dim4 &fDims)
 }
 
 template<dim_t baseDim>
-af_err fft_convolve(af_array *out, const af_array signal, const af_array filter, const bool expand)
+af_err fft_convolve(af_array *out, const af_array signal,
+                    const af_array filter, const bool expand)
 {
     try {
         const ArrayInfo& sInfo = getInfo(signal);
         const ArrayInfo& fInfo = getInfo(filter);
-
         af_dtype stype  = sInfo.getType();
-
-        dim4 sdims = sInfo.dims();
-        dim4 fdims = fInfo.dims();
+        const dim4& sdims = sInfo.dims();
+        const dim4& fdims = fInfo.dims();
 
         AF_BATCH_KIND convBT = identifyBatchKind<baseDim>(sdims, fdims);
 
         ARG_ASSERT(1, (convBT != AF_BATCH_UNSUPPORTED));
 
-        af_array output;
+#define FFTCNV(Ti, Tr, Tc, F0, F1) \
+fftConv<Ti, Tr, Tc, F0 , F1, baseDim>(signal, filter, expand, convBT)
+
+#define FALLBACK(T) \
+fftConvFallback<T , T , T , baseDim>(signal, filter, expand)
```

Comment:
You can do the same thing using some combination of type traits instead of creating macros. You don't need macros here.